PR for SRVLOGIC-209: Add the "Openshift Serverless Logic overview" assembly. (#75129)
openshift/openshift-docs: modules/serverless-logic-overview-error-handling.adoc

// Module included in the following assemblies:
// * about/serverless-logic-overview.adoc

:_content-type: CONCEPT

> [id="serverless-logic-overview-error-handling_{context}"]

Reviewer: ftirados@redhat.com would you mind take a look a this error handling section?

Author: @ftirados@redhat.com can you please review the error handling section?

Each workflow state can define error handling, which is related only to errors that may arise during its execution. Error handling defined in one state cannot be used to handle errors that happened during execution of another state during workflow execution.

Unknown errors that may arise during workflow state execution that are not explicitly handled within the workflow definition should be reported by runtime implementations and halt workflow execution.

== Error definition 

An error definition in a workflow is composed of a `name` and `code` parameters. The `name` is a short and natural language description of an error, such as `wrong parameter`. The `code` parameter helps the implementation to identify the error.

> The `code` parameter is mandatory and the engine uses different strategies to map the provided value to an exception encountered at runtime. The available strategies include fully qualified class name, error message, and status code.

Author: Added examples

// Module included in the following assemblies:
// * about/serverless-logic-overview.adoc

:_content-type: CONCEPT
[id="serverless-logic-overview-error-handling_{context}"]
= Error handling

> Openshift Serverless Logic allows you to define `explicit` error handling, meaning you can define what should happen in case of errors inside your workflow model rather than some generic error handling entity. It enables you to handle the errors that might happen during the interactions between the workflow and external systems. When an error occurs, it changes the regular workflow sequence. In such cases, a workflow state transitions to an alternative state that can potentially handle the error, instead of transitioning to the predefined state.

Reviewer 2: ```suggestion
OpenShift Serverless Logic allows you to define `explicit` error handling. You can define inside of your workflow model what should happen if errors occur, rather than relying on a generic error handling entity. Explicit error handling enables you to handle the errors that might happen during the interactions between the workflow and external systems. When an error occurs, it changes the regular workflow sequence. In these cases, a workflow state transitions to an alternative state that can potentially handle the error, instead of transitioning to the predefined state.
```

Author: Your suggestion is much better and I have incorporated it. :)

> Each workflow state can define error handling, which is related only to errors that may arise during its execution. Error handling defined in one state cannot be used to handle errors that happened during execution of another state during workflow execution.

Reviewer 2: ```suggestion
Each workflow state can define error handling, which is related only to errors that might arise during its execution. Error handling defined in one state cannot be used to handle errors that happened during execution of another state during workflow execution.
```
https://www.ibm.com/docs/en/ibm-style?topic=word-usage#m

Unknown errors that may arise during workflow state execution that are not explicitly handled within the workflow definition should be reported by runtime implementations and halt workflow execution.

> == Error definition 

Reviewer 2: I'm only going to mark this once, but all headings need an anchor ID, and headings in a module also require the "context" variable: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-titles-and-section-headings

Author: Okay! I will include this now.

Author: I have added the anchor IDs and the "context" variables.

The examples of such a definitions for both cases are as follows:

.Example of referencing a reusable JSON error definition file
[source,json]
----
{
"errors": "file://documents/reusable/errors.json"
}
----

.Example of referencing a reusable YAML error definition file
[source,yaml]
----
errors: file://documents/reusable/errors.json
----

.Example of defining a workflow errors in-line using JSON file
[source,json]
----
{
"errors": [
  {
    "name": "Service not found error",
    "code": "404",
    "description": "Server has not found anything matching the provided service endpoint information"
  }
]
}
----

.Example of defining a workflow errors in-line using YAML file
[source,yaml]
----
errors:
  - name: Service not found error
    code: '404'
    description: Server has not found anything matching the provided service endpoint
      information
----

> You can reference these defined errors by their unique name in both states `onErrors` definitions as well as in actions `nonRetryableErrors` and `retryableErrors` properties.

Author: We can remove this sentence as it is not required. It is better to remove it rather than increasing the confusion.
